### PR TITLE
admin: show more information on ddl history

### PR DESCRIPTION
**Pull request by ciscoxll**
* Add job information
  snapshot version、err、err_count.
* Add command `admin show ddl job queries xx`.
  the command query has completed job history sql max limit 10 .
 The `admin show ddl jos` command and `admin show ddl job queries job_id` use together.

### CODE REVIEW

**Comment by ciscoxll** [comment]
@zimulala @winkyao PTAL

File: parser/parser.y [hunk 1]
```diff
@@ -420,6 +421,7 @@ import (
 	cancel		"CANCEL"
 	ddl		"DDL"
 	jobs		"JOBS"
+	job		    "JOB"
```
Comment (shenli) [inline parser/parser.y]: Keep alignment

File: plan/common_plans.go [hunk 2]
```diff
@@ -43,6 +43,13 @@ type ShowDDLJobs struct {
 	baseSchemaProducer
 }
 
+// ShowDDLJobQueries is for According to job id query job query.
```
Comment (shenli) [inline plan/common_plans.go]: English syntax error

File: executor/executor.go [hunk 3]
```diff
@@ -296,6 +297,63 @@ type ShowDDLJobsExec struct {
 	jobs   []*model.Job
 }
 
+// ShowDDLJobQueriesExec represent a show DDL job queries executor.
```
Comment (shenli) [inline executor/executor.go]: represents

File: executor/executor.go [hunk 4]
```diff
@@ -296,6 +297,63 @@ type ShowDDLJobsExec struct {
...
+
+	e.jobs = append(e.jobs, jobs...)
+	e.jobs = append(e.jobs, historyJobs...)
+
+	e.cursor = 0
+	return nil
+}
+
+// Next implements the Executor Next interface.
+func (e *ShowDDLJobQueriesExec) Next(ctx context.Context) (Row, error) {
```
Comment (shenli) [inline executor/executor.go]: We could abandon the Next interface now.

Comment (ciscoxll) [reply executor/executor.go]: Has been removed.

File: parser/parser.y (hunk 1 above)
Comment (ciscoxll) [reply parser/parser.y]: ok.

**Comment by ciscoxll** [comment]
@zimulala PTAL

File: executor/executor_test.go [hunk 5]
```diff
@@ -178,6 +178,17 @@ func (s *testSuite) TestAdmin(c *C) {
 	c.Assert(row.GetString(1), Equals, historyJobs[0].State.String())
 	c.Assert(err, IsNil)
 
+	// show DDL job queries test
```
Comment (winkyao) [inline executor/executor_test.go]: please refine these tests by creating a new store.

File: executor/executor.go [hunk 6]
```diff
@@ -296,6 +297,58 @@ type ShowDDLJobsExec struct {
...
+// Open implements the Executor Open interface.
+func (e *ShowDDLJobQueriesExec) Open(ctx context.Context) error {
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	jobs, err := admin.GetDDLJobs(e.ctx.Txn())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	historyJobs, err := admin.GetHistoryDDLJobs(e.ctx.Txn())
```
Comment (zimulala) [inline executor/executor.go]: I think if the user gives the job ID, they want to get the corresponding query. So I think GetHistoryDDLJobs only return 10 jobs that cannot meet the demand.

File: executor/executor_test.go (hunk 5 above)
Comment (ciscoxll) [reply executor/executor_test.go]: already added tests.

File: executor/executor.go (hunk 6 above)
Comment (ciscoxll) [reply executor/executor.go]: I will write a document explaining that the usage code also adds a comment.

**Comment by ciscoxll** [comment]
@winkyao @zimulala PTAL

File: executor/executor.go (hunk 6 above)
Comment (shenli) [reply executor/executor.go]: Could you add a TODO here? We need to return the job that the user needs.

File: plan/planbuilder.go [hunk 7]
```diff
@@ -646,7 +650,12 @@ func buildShowDDLJobsFields() *expression.Schema {
 	schema := expression.NewSchema(make([]*expression.Column, 0, 2)...)
 	schema.Append(buildColumn("", "JOBS", mysql.TypeVarchar, 128))
 	schema.Append(buildColumn("", "STATE", mysql.TypeVarchar, 64))
+	return schema
+}
 
+func buildShowDDLJobQueriesFields() *expression.Schema {
+	schema := expression.NewSchema(make([]*expression.Column, 0, 1)...)
+	schema.Append(buildColumn("", "QUERY", mysql.TypeVarchar, 128))
```
Comment (shenli) [inline plan/planbuilder.go]: Is 128 long enough? Some MySQL clients will truncate the data according to the field length.

**Comment by shenli** [comment]
@winkyao @zimulala PTAL

File: plan/planbuilder.go (hunk 7 above)
Comment (ciscoxll) [reply plan/planbuilder.go]: Length can also be added.

File: executor/executor.go [hunk 8]
```diff
@@ -296,6 +297,61 @@ type ShowDDLJobsExec struct {
 	jobs   []*model.Job
 }
 
+// ShowDDLJobQueriesExec represents a show DDL job queries executor.
+// The `admin show ddl jos` command and `admin show ddl job queries job_id` use together.
```
Comment (shenli) [inline executor/executor.go]: What's mean by `use together`?

File: executor/executor.go [hunk 9]
```diff
@@ -296,6 +297,61 @@ type ShowDDLJobsExec struct {
...
+// The command query has completed job history sql max limit 10 .
```
Comment (shenli) [inline executor/executor.go]: Can not understand this.

**Comment by shenli** [comment]
Please fix the CI.

File: executor/executor.go (hunk 8 above)
Comment (ciscoxll) [reply executor/executor.go]: Reference to see the meaning

File: executor/executor_test.go [hunk 10]
```diff
@@ -185,6 +185,18 @@ func (s *testSuite) TestAdmin(c *C) {
...
+	tk.MustExec("use test")
+	tk.MustExec("create table admin_test2 (c1 int, c2 int, c3 int default 1, index (c1))")
+	result := tk.MustQuery(`admin show ddl job queries 43`)
```
Comment (zimulala) [inline executor/executor_test.go]: Is it unstable when we use `43`?

Comment (ciscoxll) [reply executor/executor_test.go]: New add test ddl will affect.

Comment (ciscoxll) [reply executor/executor_test.go]: New added ddl test job id incremented.
In accordance with the order and test is very stable.

File: executor/executor.go (hunk 9 above)
Comment (ciscoxll) [reply executor/executor.go]: `admin show ddl jobs`
```
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+
| JOBS                                                                                                                                                                                              | STATE  |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+
| ID:32, Type:create table, State:synced, SchemaState:public, SchemaID:1, TableID:31, RowCount:0, ArgLen:0, start time: 2018-03-13 14:34:27.314 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
| ID:30, Type:create table, State:synced, SchemaState:public, SchemaID:1, TableID:29, RowCount:0, ArgLen:0, start time: 2018-03-13 14:34:14.123 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
| ID:28, Type:create table, State:synced, SchemaState:public, SchemaID:1, TableID:27, RowCount:0, ArgLen:0, start time: 2018-03-13 14:33:50.999 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
| ID:26, Type:create table, State:synced, SchemaState:public, SchemaID:3, TableID:25, RowCount:0, ArgLen:0, start time: 2018-03-12 15:15:07.551 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
| ID:24, Type:create table, State:synced, SchemaState:public, SchemaID:3, TableID:23, RowCount:0, ArgLen:0, start time: 2018-03-12 15:15:07.423 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
| ID:22, Type:create table, State:synced, SchemaState:public, SchemaID:3, TableID:21, RowCount:0, ArgLen:0, start time: 2018-03-12 15:15:07.304 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
| ID:20, Type:create table, State:synced, SchemaState:public, SchemaID:3, TableID:19, RowCount:0, ArgLen:0, start time: 2018-03-12 15:15:07.123 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
| ID:18, Type:create table, State:synced, SchemaState:public, SchemaID:3, TableID:17, RowCount:0, ArgLen:0, start time: 2018-03-12 15:15:07.007 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
| ID:16, Type:create table, State:synced, SchemaState:public, SchemaID:3, TableID:15, RowCount:0, ArgLen:0, start time: 2018-03-12 15:15:06.896 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
| ID:14, Type:create table, State:synced, SchemaState:public, SchemaID:3, TableID:13, RowCount:0, ArgLen:0, start time: 2018-03-12 15:15:06.783 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0 | synced |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+
```
`admin show ddl job queries 26;`
```
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY                                                                                                                                                                                                                                                                                                                                                                                                  |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| CREATE TABLE IF NOT EXISTS mysql.gc_delete_range (
		job_id BIGINT NOT NULL COMMENT "the DDL job ID",
		element_id BIGINT NOT NULL COMMENT "the schema element ID",
		start_key VARCHAR(255) NOT NULL COMMENT "encoded in hex",
		end_key VARCHAR(255) NOT NULL COMMENT "encoded in hex",
		ts BIGINT NOT NULL COMMENT "timestamp in int64",
		UNIQUE KEY (element_id),
		KEY (job_id, element_id)
	); |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

File: executor/executor.go (hunk 8 above)
Comment (shenli) [reply executor/executor.go]: are used together.

File: executor/executor.go (hunk 9 above)
Comment (shenli) [reply executor/executor.go]: The comment is not clear.

**Comment by ciscoxll** [comment]
@zimulala @winkyao @ PTAL

File: executor/executor.go [hunk 11]
```diff
@@ -296,6 +297,60 @@ type ShowDDLJobsExec struct {
 	jobs   []*model.Job
 }
 
+// ShowDDLJobQueriesExec represents a show DDL job queries executor.
+// The command query has completed job history sql limit 10.
```
Comment (winkyao) [inline executor/executor.go]: The jobs id that is given by 'admin show ddl job queries' statement, only be searched in the latest 10 history jobs.

File: executor/executor.go [hunk 12]
```diff
@@ -296,6 +297,60 @@ type ShowDDLJobsExec struct {
...
+	// TODO: need to return the job that the user needs.
+	historyJobs, err := admin.GetHistoryDDLJobs(e.ctx.Txn())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	e.jobs = append(e.jobs, jobs...)
+	e.jobs = append(e.jobs, historyJobs...)
+
+	e.cursor = 0
```
Comment (zimulala) [inline executor/executor.go]: Could we remove this code？ The original value is 0.

File: executor/executor.go (hunk 11 above)
Comment (ciscoxll) [reply executor/executor.go]: That's what it means.

File: executor/executor_test.go (hunk 5 above)
Comment (zimulala) [reply executor/executor_test.go]: Where did we use a new store in this test?
If we don't use a new store, I think the result in line 193 will be unstable.

**Comment by shenli** [comment]
@ciscoxll Please fix the CI.

File: executor/executor.go (hunk 12 above)
Comment (ciscoxll) [reply executor/executor.go]: ok

**Comment by ciscoxll** [comment]
@zimulala @winkyao @ PTAL

File: executor/executor_test.go [hunk 13]
```diff
@@ -185,6 +185,20 @@ func (s *testSuite) TestAdmin(c *C) {
 	c.Assert(row.GetString(1), Equals, historyJobs[0].State.String())
 	c.Assert(err, IsNil)
 
+	// show DDL job queries test
+	s.TearDownSuite(c)
+	s.SetUpSuite(c)
```
Comment (zimulala) [inline executor/executor_test.go]: We'd better not use `TearDownSuite` and `SetUpSuite` in a specific test.
Rest LGTM

**Comment by ciscoxll** [comment]
@zimulala @winkyao @ PTAL

**Comment by zimulala** [comment]
LGTM

**Comment by ciscoxll** [comment]
@winkyao LGTM

**Comment by zimulala** [comment]
Please resolve the confliction.

**Pull request merged**